capture/src/v0_endpoint.rs

```diff
@@ -86,41 +89,36 @@ pub async fn event(
                     tracing::error!("failed to decode form data: {}", e);
                     CaptureError::RequestDecodingError(String::from("missing data field"))
                 })?;
-            RawEvent::from_bytes(payload.into())
+            RawRequest::from_bytes(payload.into())
         }
         ct => {
             tracing::Span::current().record("content_type", ct);
 
-            RawEvent::from_bytes(body)
+            RawRequest::from_bytes(body)
         }
     }?;
 
+    let sent_at = request.sent_at().or(meta.sent_at());
+    let token = match request.extract_and_verify_token() {
+        Ok(token) => token,
+        Err(err) => {
+            report_dropped_events("token_shape_invalid", request.events().len() as u64);
+            return Err(err);
+        }
+    };
```

Comment:
If I'm reading this correctly, for the `/batch` variant, `extract_and_verify_token` will look for tokens in each of the events in the batch. However, the token may also be present at the top level, and omitted from each event in the batch.

Maybe I'm missing something about how we would extract the token in that case?

Reply:
On the `/batch` shape (the `RawRequest::Batch` enum variant), the token is taken from the mandatory toplevel field. The code to scan all events is for the `RawRequest::Array` shape (that's the posthog-js shape with an array of events and no toplevel metadata).

python will most probably fallback to scanning events if the toplevel token is not given, but the double-run will tell us whether this fallback is needed.

Reply:
Ah, I see! Looks good! I wasn't able to find the match for some reason, I think due to the file being moved I was being sent to another function with the same name.

Reply:
Thank you!

Reply:
Right now if the toplevel `token` (or `api_key` as alias) field is not present, the request will 400 with a `RequestParsingError` (no variant matching) that'll show in traces. If we want to allow falling-back to token in events we'd need to make token and `Option<String>` for serde to not error out.